The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5701771761?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/e9f08e709d471988fcf163779dec430cdcbf1193/checks/5701771761/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] codegen/loads.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-12/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/loads/loads.ll" "/checkout/src/test/codegen/loads.rs" "--check-prefixes" "CHECK,NONMSVC"
stdout: none
--- stderr -------------------------------
/checkout/src/test/codegen/loads.rs:125:11: error: CHECK: expected string not found in input
// CHECK: load {{(i32\*, )?}}i32** %x{{.*}}, !nonnull
          ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/loads/loads.ll:612:17: note: scanning from here
define i32 @_box(i32* noalias noundef nonnull align 4 %0) unnamed_addr #1 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality {
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/loads/loads.ll:617:7: note: possible intended match here
 %2 = load i32*, i32** %x, align 8

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/loads/loads.ll
Check file: /checkout/src/test/codegen/loads.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
             .
             .
             .
             .
           607:  %1 = load i32*, i32** %x, align 8, !nonnull !3, !align !4, !noundef !3
           608:  ret i32* %1
           609: }
           610: 
           611: ; Function Attrs: nonlazybind uwtable
           612: define i32 @_box(i32* noalias noundef nonnull align 4 %0) unnamed_addr #1 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality {
check:125'0                     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           613: start:
check:125'0     ~~~~~~
           614:  %1 = alloca { i8*, i32 }, align 8
check:125'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           615:  %x = alloca i32*, align 8
check:125'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
           616:  store i32* %0, i32** %x, align 8
check:125'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           617:  %2 = load i32*, i32** %x, align 8
check:125'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:125'1           ?                            possible intended match
           618:  %3 = load i32, i32* %2, align 4
check:125'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           619: ; invoke core::ptr::drop_in_place<alloc::boxed::Box<i32>>
check:125'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           620:  invoke void @"_ZN4core3ptr49drop_in_place$LT$alloc..boxed..Box$LT$i32$GT$$GT$17h78d62f8a4e32a77eE"(i32** %x)
check:125'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           621:  to label %bb1 unwind label %cleanup
check:125'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           622: 
check:125'0     ~
             .
             .
>>>>>>
------------------------------------------
```

</details>

--- a/compiler/rustc_codegen_ssa/src/mir/place.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/place.rs
@@ -441,11 +441,19 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                     .find(|elem| matches!(elem.1, mir::ProjectionElem::Deref))
                 {
                     base = elem.0 + 1;
-                    self.codegen_consume(
+                    let cg_base = self.codegen_consume(
                         bx,
                         mir::PlaceRef { projection: &place_ref.projection[..elem.0], ..place_ref },
-                    )
-                    .deref(bx.cx())
+                    );
+
+                    // a box with a non-zst allocator should not be directly dereferenced
+                    if cg_base.layout.ty.is_box() && !cg_base.layout.field(cx, 1).is_zst() {
+                        let ptr = cg_base.extract_field(bx, 0).extract_field(bx, 0);
+
+                        ptr.deref(bx.cx())
+                    } else {
+                        cg_base.deref(bx.cx())
+                    }
                 } else {
                     bug!("using operand local {:?} as place", place_ref);
                 }

--- a/compiler/rustc_codegen_ssa/src/mir/place.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/place.rs
@@ -455,7 +455,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
             cg_base = match elem.clone() {
                 mir::ProjectionElem::Deref => {
                     // a box with a non-zst allocator should not be directly dereferenced
-                    if cg_base.layout.ty.is_box() && !cg_base.layout.field(cx, 0).is_zst() {
+                    if cg_base.layout.ty.is_box() && !cg_base.layout.field(cx, 1).is_zst() {
                         let ptr = cg_base.project_field(bx, 0).project_field(bx, 0);
 
                         bx.load_operand(ptr).deref(bx.cx())

--- a/src/test/ui/box/issue-95036.rs
+++ b/src/test/ui/box/issue-95036.rs
@@ -3,8 +3,20 @@
 
 #![feature(allocator_api, bench_black_box)]
 
+#[inline(never)]
+pub fn by_ref(node: &mut Box<[u8; 1], &std::alloc::Global>) {
+    node[0] = 9u8;
+}
+
 pub fn main() {
     let mut node = Box::new_in([5u8], &std::alloc::Global);
     node[0] = 7u8;
+
+    std::hint::black_box(node);
+
+    let mut node = Box::new_in([5u8], &std::alloc::Global);
+
+    by_ref(&mut node);
+
     std::hint::black_box(node);
 }

